allennlp/modules/elmo.py

```diff
@@ -0,0 +1,422 @@
+        return self.output_dim
+
+    @overrides
+    def forward(self, inputs: torch.Tensor) -> Dict[str, torch.Tensor]:  # pylint: disable=arguments-differ
+        """
+        Compute context insensitive token embeddings for ELMo representations.
+
+        Parameters
+        ----------
+        inputs: ``torch.autograd.Variable``
+            Shape ``(batch_size, sequence_length, 50)`` of character ids representing the
+            current batch.
+
+        Returns
+        -------
+        Dict with keys:
+
+        ``'token_embedding'``: ``torch.autograd.Variable``
+            Shape ``(batch_size, sequence_length + 2, embedding_dim)`` tensor with context
+            insensitive token representations.
+        ``'mask'``:  ``torch.autograd.Variable``
+            Shape ``(batch_size, sequence_length + 2)`` long tensor with sequence mask.
+        """
+        # Add BOS/EOS
+        mask = ((inputs > 0).long().sum(dim=-1) > 0).long()
+        character_ids_with_bos_eos, mask_with_bos_eos = add_sentence_boundary_token_ids(
+                inputs,
+                mask,
+                self._beginning_of_sentence_characters,
+                self._end_of_sentence_characters
+        )
+
+        # the character id embedding
+        max_chars_per_token = self._options['char_cnn']['max_characters_per_token']
+        # (batch_size * sequence_length, max_chars_per_token, embed_dim)
+        character_embedding = torch.nn.functional.embedding(
+                character_ids_with_bos_eos.view(-1, max_chars_per_token),
+                self._char_embedding_weights
+        )
+
+        # run convolutions
+        cnn_options = self._options['char_cnn']
+        if cnn_options['activation'] == 'tanh':
+            activation = torch.nn.functional.tanh
+        elif cnn_options['activation'] == 'relu':
+            activation = torch.nn.functional.relu
+        else:
+            raise ConfigurationError("Unknown activation")
+
+        # (batch_size * sequence_length, embed_dim, max_chars_per_token)
+        character_embedding = torch.transpose(character_embedding, 1, 2)
+        convs = []
+        for conv in self._convolutions:
+            convolved = conv(character_embedding)
+            # (batch_size * sequence_length, n_filters for this width)
+            convolved, _ = torch.max(convolved, dim=-1)
+            convolved = activation(convolved)
+            convs.append(convolved)
+
+        # (batch_size * sequence_length, n_filters)
+        token_embedding = torch.cat(convs, dim=-1)
+
+        # apply the highway layers (batch_size * sequence_length, n_filters)
+        token_embedding = self._highways(token_embedding)
+
+        # final projection  (batch_size * sequence_length, embedding_dim)
+        token_embedding = self._projection(token_embedding)
+
+        # reshape to (batch_size, sequence_length, embedding_dim)
+        batch_size, sequence_length, _ = character_ids_with_bos_eos.size()
+
+        return {
```

Comment:
I would mildly prefer that these return tuples, just because 1) There's only 2 arguments and 2) All of the other modules return tuples, so it's a little surprising. I appreciate the reasons, so if you feel strongly, fine.

Reply:
I'd rather keep them a dicts, makes changing in backward compatible manner easier.  And the return object is self documenting via the keys.